js/ui/data_grid.d.ts

```diff
@@ -69,767 +69,1019 @@ import Widget, {
+  readonly model?: any;
+  readonly key: any;
+  cancel: boolean;
+}
+
+/**
+ * @public
+ */
+export interface RowInsertedEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly data: any;
+  readonly key: any;
+  readonly error: Error;
+}
+
+/**
+ * @public
+ */
+export interface RowInsertingEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly data: any;
+  cancel: boolean | TPromise<void>;
+}
+
+/**
+ * @public
+ */
+export interface RowRemovedEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly data: any;
+  readonly key: any;
+  readonly error: Error;
+}
+
+/**
+ * @public
+ */
+export interface RowRemovingEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly data: any;
+  readonly key: any;
+  cancel: boolean | TPromise<void>;
+}
+
+/**
+ * @public
+ */
+export interface RowUpdatedEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly data: any;
+  readonly key: any;
+  readonly error: Error;
+}
+
+/**
+ * @public
+ */
+export interface RowUpdatingEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly oldData: any;
+  newData: any;
+  readonly key: any;
+  cancel: boolean | TPromise<void>;
+}
+
+/**
+ * @public
+ */
+export interface RowValidatingEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly brokenRules: Array<RequiredRule | NumericRule | RangeRule | StringLengthRule | CustomRule | CompareRule | PatternRule | EmailRule | AsyncRule>;
+  isValid: boolean;
+  readonly key: any;
+  readonly newData: any;
+  readonly oldData: any;
+  errorText: string;
+  promise?: TPromise<void>;
+}
+
+/**
+ * @public
+ */
+export interface SavedEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly changes: Array<any>;
+}
+
+
+/**
+ * @public
+ */
+export interface SavingEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  changes: Array<any>;
+  promise?: TPromise<void>;
+  cancel: boolean;
+}
+
+/**
+ * @public
+ */
+export interface SelectionChangedEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  readonly currentSelectedRowKeys: Array<any>;
+  readonly currentDeselectedRowKeys: Array<any>;
+  readonly selectedRowKeys: Array<any>;
+  readonly selectedRowsData: Array<any>;
+}
+
+/**
+ * @public
+ */
+export interface ToolbarPreparingEvent<T extends GridBase> {
+  readonly component: T;
+  readonly element: TElement;
+  readonly model?: any;
+  toolbarOptions: dxToolbarOptions;
```

Comment:
But by code, it is possible to modify it and it will work. We can discuss it.

Reply:
Again, it's look like we need to discuss these cases and make one decision for all of them. Maybe with some restrictions but still.

Reply:
Or postpone this discussion for a better time)

Reply:
I don't remember that customers change the entire toolbarOptions object. Usually, they modify items. However, there may be cases that customers did it and did not ask any questions, because it worked.